-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-82d89f/26843/summary.html
**COMMIT**: a091b017f2569d7ee0f4a35abfe9d32354d3d36a
**CMSSW**: CMSSW_12_5_X_2022-08-15-2300/el8_amd64_gcc10
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38877/26843/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>Entering library rule at src/HeterogeneousCore/AlpakaTest/plugins
>> Compiling alpaka/serial /pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-15-2300/src/HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlgo.dev.cc
>> Compiling alpaka/serial edm plugin /pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-15-2300/src/HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaProducer.cc
>> Compiling alpaka/serial edm plugin /pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-15-2300/src/HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaTranscriber.cc
/pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-15-2300/src/HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaTranscriber.cc: In member function 'virtual void alpaka_serial_sync::TestAlpakaTranscriber::acquire(const edm::Event&, const edm::EventSetup&, edm::WaitingTaskWithArenaHolder)':
/pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-15-2300/src/HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaTranscriber.cc:46:100: error: use of deleted function 'PortableHostCollection<T>& PortableHostCollection<T>::operator=(PortableHostCollection<T>&&) [with T = portabletest::TestSoALayout<>]'
   46 |       hostProduct_ = portabletest::TestHostCollection{deviceProduct->metadata().size(), ctx.queue()};
      |                                                                                                    ^
In file included from /pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-15-2300/src/DataFormats/PortableTestObjects/interface/TestHostCollection.h:4,
                 from /pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-15-2300/src/HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaTranscriber.cc:7:
/cvmfs/cms-ib.cern.ch/nweek-02746/el8_amd64_gcc10/cms/cmssw/CMSSW_12_5_X_2022-08-15-2300/src/DataFormats/Portable/interface/PortableHostCollection.h:51:27: note: 'PortableHostCollection<T>& PortableHostCollection<T>::operator=(PortableHostCollection<T>&&) [with T = portabletest::TestSoALayout<>]' is implicitly deleted because the default definition would be ill-formed:
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 64 COMPILER='llvm compile'

```


<pre>>> Creating project symlinks
>> Entering Package HeterogeneousCore/AlpakaTest
>> Entering Package DataFormats/SoATemplate
>> Entering Package DataFormats/PortableTestObjects
>> Compile sequence completed for CMSSW CMSSW_12_5_X_2022-08-15-2300
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
+ eval scram build outputlog '&&' '(python3' /pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/cms-bot/buildLogAnalyzer.py --ignoreWarning=Wdeprecated-declarations --logDir /pool/condor/dir_19852/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-15-2300/tmp/el8_amd64_gcc10/cache/log/src '||' 'true)'
++ scram build outputlog
>> Entering Package DataFormats/SoATemplate
Entering library rule at src/DataFormats/SoATemplate/test
>> Building LCG reflex dict from header file src/DataFormats/SoATemplate/test/classes.h
</pre>

--- a/DataFormats/SoATemplate/interface/SoALayout.h
+++ b/DataFormats/SoATemplate/interface/SoALayout.h
@@ -491,9 +491,37 @@
       const CLASS& parent_;                                                                                            \
       using ParentClass = CLASS;                                                                                       \
     };                                                                                                                 \
+                                                                                                                       \
     friend Metadata;                                                                                                   \
+                                                                                                                       \
     SOA_HOST_DEVICE SOA_INLINE const Metadata metadata() const { return Metadata(*this); }                             \
     SOA_HOST_DEVICE SOA_INLINE Metadata metadata() { return Metadata(*this); }                                         \
+                                                                                                                       \
+    /* Generate the ConstView template */                                                                              \
+    _GENERATE_SOA_TRIVIAL_CONST_VIEW(CLASS,                                                                            \
+                    SOA_VIEW_LAYOUT_LIST(                                                                              \
+                        SOA_VIEW_LAYOUT(BOOST_PP_CAT(CLASS, _parametrized) , BOOST_PP_CAT(instance_, CLASS))),         \
+                    SOA_VIEW_VALUE_LIST(_ITERATE_ON_ALL_COMMA(                                                         \
+                    _VIEW_FIELD_FROM_LAYOUT, BOOST_PP_CAT(instance_, CLASS), __VA_ARGS__)))                            \
+                                                                                                                       \
+    template <bool RESTRICT_QUALIFY, bool RANGE_CHECKING>                                                              \
+    using ConstViewTemplate = ConstViewTemplateFreeParams<ALIGNMENT, ALIGNMENT_ENFORCEMENT, RESTRICT_QUALIFY,          \
+      RANGE_CHECKING>;                                                                                                 \
+                                                                                                                       \
+    using ConstView = ConstViewTemplate<cms::soa::RestrictQualify::enabled, cms::soa::RangeChecking::disabled>;        \
+                                                                                                                       \
+    /* Generate the mutable View template */                                                                           \
+    _GENERATE_SOA_TRIVIAL_VIEW(CLASS,                                                                                  \
+                    SOA_VIEW_LAYOUT_LIST(                                                                              \
+                        SOA_VIEW_LAYOUT(BOOST_PP_CAT(CLASS, _parametrized), BOOST_PP_CAT(instance_, CLASS))),          \
+                    SOA_VIEW_VALUE_LIST(_ITERATE_ON_ALL_COMMA(                                                         \
+                    _VIEW_FIELD_FROM_LAYOUT, BOOST_PP_CAT(instance_, CLASS), __VA_ARGS__)),                            \
+                    __VA_ARGS__)                                                                                       \
+                                                                                                                       \
+    template <bool RESTRICT_QUALIFY, bool RANGE_CHECKING>                                                              \
+    using ViewTemplate = ViewTemplateFreeParams<ALIGNMENT, ALIGNMENT_ENFORCEMENT, RESTRICT_QUALIFY, RANGE_CHECKING>;   \
+                                                                                                                       \
+    using View = ViewTemplate<cms::soa::RestrictQualify::enabled, cms::soa::RangeChecking::disabled>;                  \
                                                                                                                        \
     /* Trivial constuctor */                                                                                           \
     CLASS()                                                                                                            \
@@ -522,7 +550,19 @@
         return *this;                                                                                                  \
     }                                                                                                                  \
                                                                                                                        \
+    /* ROOT read streamer */                                                                                           \
+    template <typename T>                                                                                              \
+    void ROOTReadStreamer(T & onfile) {                                                                                \
+      auto size = onfile.metadata().size();                                                                            \
+      _ITERATE_ON_ALL(_STREAMER_READ_SOA_DATA_MEMBER, ~, __VA_ARGS__)                                                  \
+    }                                                                                                                  \
+                                                                                                                       \
+    /* Dump the SoA internal structure */                                                                              \
+    template <typename T>                                                                                              \
+    SOA_HOST_ONLY friend void dump();                                                                                  \
+                                                                                                                       \
   private:                                                                                                             \
+    /* Helper method for the user provided storage constructor and ROOT streamer */                                    \
     void organizeColumnsFromBuffer() {                                                                                 \
       if constexpr (alignmentEnforcement == cms::soa::AlignmentEnforcement::enforced)                                  \
         if (reinterpret_cast<intptr_t>(mem_) % alignment)                                                              \
@@ -535,19 +575,6 @@
         throw std::runtime_error("In " #CLASS "::" #CLASS ": unexpected end pointer.");                                \
     }                                                                                                                  \
                                                                                                                        \
-  public:                                                                                                              \
-    /* ROOT read streamer */                                                                                           \
-    template <typename T>                                                                                              \
-    void ROOTReadStreamer(T & onfile) {                                                                                \
-      auto size = onfile.metadata().size();                                                                            \
-      _ITERATE_ON_ALL(_STREAMER_READ_SOA_DATA_MEMBER, ~, __VA_ARGS__)                                                  \
-    }                                                                                                                  \
-                                                                                                                       \
-    /* dump the SoA internal structure */                                                                              \
-    template <typename T>                                                                                              \
-    SOA_HOST_ONLY friend void dump();                                                                                  \
-                                                                                                                       \
-  private:                                                                                                             \
     /* Range checker conditional to the macro _DO_RANGECHECK */                                                        \
     SOA_HOST_DEVICE SOA_INLINE                                                                                         \
     void rangeCheck(size_type index) const {                                                                           \
@@ -559,7 +586,7 @@
       }                                                                                                                \
     }                                                                                                                  \
                                                                                                                        \
-    /* data members */                                                                                                 \
+    /* Data members */                                                                                                 \
     std::byte* mem_;                                                                                                   \
     size_type elements_;                                                                                               \
     size_type const scalar_ = 1;                                                                                       \
@@ -569,29 +596,6 @@
     /* So instead we make the code unconditional with paceholder names which are protected by a private protection. */ \
     /* This will be handled later as we handle the integration of the view as a subclass of the layout.             */ \
                                                                                                                        \
-  public:                                                                                                              \
-  _GENERATE_SOA_TRIVIAL_CONST_VIEW(CLASS,                                                                              \
-                    SOA_VIEW_LAYOUT_LIST(                                                                              \
-                        SOA_VIEW_LAYOUT(BOOST_PP_CAT(CLASS, _parametrized) , BOOST_PP_CAT(instance_, CLASS))),         \
-                    SOA_VIEW_VALUE_LIST(_ITERATE_ON_ALL_COMMA(                                                         \
-                    _VIEW_FIELD_FROM_LAYOUT, BOOST_PP_CAT(instance_, CLASS), __VA_ARGS__)))                            \
-                                                                                                                       \
-    template <bool RESTRICT_QUALIFY, bool RANGE_CHECKING>                                                              \
-    using ConstViewTemplate = ConstViewTemplateFreeParams<ALIGNMENT, ALIGNMENT_ENFORCEMENT, RESTRICT_QUALIFY,          \
-      RANGE_CHECKING>;                                                                                                 \
-                                                                                                                       \
-  _GENERATE_SOA_TRIVIAL_VIEW(CLASS,                                                                                    \
-                    SOA_VIEW_LAYOUT_LIST(                                                                              \
-                        SOA_VIEW_LAYOUT(BOOST_PP_CAT(CLASS, _parametrized), BOOST_PP_CAT(instance_, CLASS))),          \
-                    SOA_VIEW_VALUE_LIST(_ITERATE_ON_ALL_COMMA(                                                         \
-                    _VIEW_FIELD_FROM_LAYOUT, BOOST_PP_CAT(instance_, CLASS), __VA_ARGS__)),                            \
-                    __VA_ARGS__)                                                                                       \
-                                                                                                                       \
-    template <bool RESTRICT_QUALIFY, bool RANGE_CHECKING>                                                              \
-    using ViewTemplate = ViewTemplateFreeParams<ALIGNMENT, ALIGNMENT_ENFORCEMENT, RESTRICT_QUALIFY, RANGE_CHECKING>;   \
-                                                                                                                       \
-    using ConstView = ConstViewTemplate<cms::soa::RestrictQualify::enabled, cms::soa::RangeChecking::disabled>;        \
-    using View = ViewTemplate<cms::soa::RestrictQualify::enabled, cms::soa::RangeChecking::disabled>;                  \
   };
 // clang-format on
 

--- a/DataFormats/SoATemplate/interface/SoALayout.h
+++ b/DataFormats/SoATemplate/interface/SoALayout.h
@@ -181,6 +181,37 @@
 #define _DECLARE_MEMBER_TRIVIAL_CONSTRUCTION(R, DATA, TYPE_NAME) \
   BOOST_PP_EXPAND(_DECLARE_MEMBER_TRIVIAL_CONSTRUCTION_IMPL TYPE_NAME)
 
+// clang-format off
+#define _DECLARE_MEMBER_COPY_CONSTRUCTION_IMPL(VALUE_TYPE, CPP_TYPE, NAME)                                             \
+  _SWITCH_ON_TYPE(VALUE_TYPE,                                                                                          \
+      /* Scalar */                                                                                                     \
+      (BOOST_PP_CAT(NAME, _){other.BOOST_PP_CAT(NAME, _)}),                                                            \
+      /* Column */                                                                                                     \
+      (BOOST_PP_CAT(NAME, _){other.BOOST_PP_CAT(NAME, _)}),                                                            \
+      /* Eigen column */                                                                                               \
+      (BOOST_PP_CAT(NAME, _){other.BOOST_PP_CAT(NAME, _)})                                                             \
+      (BOOST_PP_CAT(NAME, Stride_){other.BOOST_PP_CAT(NAME, Stride_)})                                                 \
+  )
+// clang-format on
+
+#define _DECLARE_MEMBER_COPY_CONSTRUCTION(R, DATA, TYPE_NAME) \
+  BOOST_PP_EXPAND(_DECLARE_MEMBER_COPY_CONSTRUCTION_IMPL TYPE_NAME)
+
+// clang-format off
+#define _DECLARE_MEMBER_ASSIGNMENT_IMPL(VALUE_TYPE, CPP_TYPE, NAME)                                                    \
+  _SWITCH_ON_TYPE(VALUE_TYPE,                                                                                          \
+      /* Scalar */                                                                                                     \
+      BOOST_PP_CAT(NAME, _) = other.BOOST_PP_CAT(NAME, _);,                                                          \
+      /* Column */                                                                                                     \
+      BOOST_PP_CAT(NAME, _) = other.BOOST_PP_CAT(NAME, _);,                                                          \
+      /* Eigen column */                                                                                               \
+      BOOST_PP_CAT(NAME, _) = other.BOOST_PP_CAT(NAME, _);                                                           \
+      BOOST_PP_CAT(NAME, Stride_) = other.BOOST_PP_CAT(NAME, Stride_);                                               \
+  )
+// clang-format on
+
+#define _DECLARE_MEMBER_ASSIGNMENT(R, DATA, TYPE_NAME) BOOST_PP_EXPAND(_DECLARE_MEMBER_ASSIGNMENT_IMPL TYPE_NAME)
+
 /**
  * Declare the value_element data members
  */
@@ -474,6 +505,21 @@
     /* Constructor relying on user provided storage (implementation shared with ROOT streamer) */                      \
     SOA_HOST_ONLY CLASS(std::byte* mem, size_type elements) : mem_(mem), elements_(elements), byteSize_(0) {           \
       organizeColumnsFromBuffer();                                                                                     \
+    }                                                                                                                  \
+                                                                                                                       \
+    /* Explicit copy constructor and assignment operator */                                                            \
+    SOA_HOST_ONLY CLASS(CLASS const& other)                                                                            \
+        : mem_(other.mem_),                                                                                            \
+          elements_(other.elements_),                                                                                  \
+          byteSize_(other.byteSize_),                                                                                  \
+          _ITERATE_ON_ALL_COMMA(_DECLARE_MEMBER_COPY_CONSTRUCTION, ~, __VA_ARGS__) {}                                  \
+                                                                                                                       \
+    SOA_HOST_ONLY CLASS& operator=(CLASS const& other) {                                                               \
+        mem_ = other.mem_;                                                                                             \
+        elements_ = other.elements_;                                                                                   \
+        byteSize_ = other.byteSize_;                                                                                   \
+        _ITERATE_ON_ALL(_DECLARE_MEMBER_ASSIGNMENT, ~, __VA_ARGS__)                                                    \
+        return *this;                                                                                                  \
     }                                                                                                                  \
                                                                                                                        \
   private:                                                                                                             \

--- a/DataFormats/SoATemplate/interface/SoALayout.h
+++ b/DataFormats/SoATemplate/interface/SoALayout.h
@@ -536,15 +536,6 @@
     }                                                                                                                  \
                                                                                                                        \
   public:                                                                                                              \
-    /* Constructor relying on user provided storage */                                                                 \
-    SOA_DEVICE_ONLY CLASS(bool devConstructor, std::byte* mem, size_type elements) :                                   \
-      mem_(mem),                                                                                                       \
-      elements_(elements)                                                                                              \
-    {                                                                                                                  \
-      auto curMem = mem_;                                                                                              \
-      _ITERATE_ON_ALL(_ASSIGN_SOA_COLUMN_OR_SCALAR, ~, __VA_ARGS__)                                                    \
-    }                                                                                                                  \
-                                                                                                                       \
     /* ROOT read streamer */                                                                                           \
     template <typename T>                                                                                              \
     void ROOTReadStreamer(T & onfile) {                                                                                \
@@ -577,6 +568,7 @@
     /* Making the code conditional is problematic in macros as the commas will interfere with parameter lisings     */ \
     /* So instead we make the code unconditional with paceholder names which are protected by a private protection. */ \
     /* This will be handled later as we handle the integration of the view as a subclass of the layout.             */ \
+                                                                                                                       \
   public:                                                                                                              \
   _GENERATE_SOA_TRIVIAL_CONST_VIEW(CLASS,                                                                              \
                     SOA_VIEW_LAYOUT_LIST(                                                                              \

